runtime/dart_isolate.cc

```diff
@@ -274,35 +274,50 @@ bool DartIsolate::PrepareForRunningFromPrecompiledCode() {
   return true;
 }
 
-static bool LoadScriptSnapshot(std::shared_ptr<const fml::Mapping> mapping) {
+static bool LoadScriptSnapshot(std::shared_ptr<const fml::Mapping> mapping,
+                               bool last_piece) {
+  FXL_CHECK(last_piece) << "Script snapshots cannot be divided";
   if (tonic::LogIfError(Dart_LoadScriptFromSnapshot(mapping->GetMapping(),
                                                     mapping->GetSize()))) {
     return false;
   }
   return true;
 }
 
-static bool LoadKernelSnapshot(std::shared_ptr<const fml::Mapping> mapping) {
-  if (tonic::LogIfError(Dart_LoadScriptFromKernel(mapping->GetMapping(),
-                                                  mapping->GetSize()))) {
+static bool LoadKernelSnapshot(std::shared_ptr<const fml::Mapping> mapping,
+                               bool last_piece) {
+  Dart_Handle library = Dart_LoadLibraryFromKernel(mapping->GetMapping(),
+                                                   mapping->GetSize());
+  if (tonic::LogIfError(library)) {
     return false;
   }
 
+  if (!last_piece) {
+    // More to come.
+    return true;
+  }
+
+  Dart_SetRootLibrary(library);
```

Comment:
The last piece might not necessarily be the root library. Maybe add a TODO here and other places where changes will be needed to handle out-of-order/async loading.

Reply:
runtime/dart_runner/kernel/compiler.dart ensures the piece with main is the last piece. We could ensure whoever issues the async loads presents the last one last, or we could have separate "main" and "done" bits.